portality/scripts/article_duplicates_report_remove.py

```diff
@@ -0,0 +1,139 @@
+"""
+This script will run through the entire catalogue, detecting duplicates
+of articles and removing them. It could take a long time and you should
+take great care running it - at minimum double check there is a recent
+backup available, there should always be one anyway.
+"""
+
+from portality.models import Article, Journal, Account
+from portality.article import XWalk
+from portality.core import app
+from datetime import datetime
+import time
+import esprit
+
+
+def duplicates_per_article(connection, delete, snapshot, owner=None, query_override=None):
+    """Scroll through all articles, finding (and deleting if set) duplicates for each."""
+    dupcount = 0
+    deleted_ids = set()
+
+    if query_override:
+        scroll_query = query_override
+    else:
+        scroll_query = {                                                                  # scroll from newest to oldest
+            "query": {"match_all": {}},
+            "sort": [{"last_updated": {"order": "desc"}}]
+        }
+
+    for a in esprit.tasks.scroll(connection, 'article', q=scroll_query):
+        article = Article(_source=a)
+
+        if article.id in deleted_ids:                               # only delete its duplicates if it isn't a duplicate
+            continue
+
+        duplicates = XWalk.discover_duplicates(article, owner)
+        if duplicates:
+            doi_dups = [d.id for d in duplicates.get('doi', [])]
+            fulltext_dups = [d.id for d in duplicates.get('fulltext', [])]
+            doi_dups.sort(), fulltext_dups.sort()                            # Sort so we can visually compare the lists
+            print "\n{0}".format(article.id)
+            if doi_dups:
+                print "\t{0} DOI duplicates: {1}".format(len(doi_dups), ", ".join(doi_dups))
+            if fulltext_dups:
+                print "\t{0} fulltext duplicates: {1}".format(len(fulltext_dups), ", ".join(fulltext_dups))
+
+            set_of_duplicates = set(doi_dups + fulltext_dups)
+            dupcount += len(set_of_duplicates) + 1                     # The detected duplicates plus the article itself
+
+            if delete:
+                for dup in set_of_duplicates:
+                    dup_art = article.pull(dup)
+                    if dup_art is not None:
+                        if snapshot:
+                            dup_art.snapshot()
+                        dup_art.delete()
+                        deleted_ids.add(dup)
+            else:
+                deleted_ids = deleted_ids | set_of_duplicates
+
+    return dupcount, len(deleted_ids)
+
+
+def duplicates_per_account(connection, delete, snapshot):
+    dupcount = 0
+    delcount = 0
+
+    for acc in esprit.tasks.scroll(connection, 'account'):
+        account = Account(**acc)
+
+        js = Journal.all(terms={"admin.owner": [account.id]})
```

Comment:
Probably needs to be admin.owner.exact.  Might also be worth pushing these behaviours into the Journal object.  See Journal.issns_by_owner for a possible template for this.